lusidtools/cocoon/cocoon.py

```diff
@@ -901,11 +904,342 @@ async def _construct_batches(
+            code_tuples.append((code, effective_at))
+    return list(set(code_tuples))
+
+
+def return_unmatched_transactions(
+    api_factory: lusid.utilities.ApiClientFactory, scope: str, code: str
+):
+    """
+    Call the get transactions api and only return those transactions with unresolved identifiers.
+
+    Multiple pages of responses for the get_transactions api call are handled.
+
+    Parameters
+    ----------
+    api_factory : lusid.utilities.ApiClientFactory api_factory
+        The api factory to use
+    scope : str
+        The scope of the resource to load the data into
+    code : str
+        The code of the portfolio containing the transactions we want to return
+
+    Returns
+    -------
+    A list of objects with the structure { transaction_id: instrument_identifiers }.
+    """
+    transactions_api = api_factory.build(lusid.api.TransactionPortfoliosApi)
+    done = False
+    next_page = None
+    unmatched_transactions = []
+
+    # We need to handle the possibility of paginated results
+    while not done:
+
+        # There must be a filter included so only transactions with unmatched instruments are returned
+        kwargs = {
+            "scope": scope,
+            "code": code,
+            "filter": "instrumentUid eq'LUID_ZZZZZZZZ'",
+        }
+
+        if next_page is not None:
+            kwargs["page"] = next_page
+
+        response = transactions_api.get_transactions(**kwargs)
+
+        unmatched_transactions.extend(
+            [
+                {response.transaction_id: response.instrument_identifiers}
+                for response in response.values
+            ]
+        )
+
+        next_page = response.next_page
+        done = response.next_page is None
+
+    return unmatched_transactions
+
+
+def filter_unmatched_transactions(
+    data_frame: pd.DataFrame, mapping_required: dict, unmatched_transactions: list,
+):
+    """
+    This method will take the full list of unmatched transactions and remove any transactions that were not
+    part of the current upload with load_from_data_frame.
+
+    Parameters
+    ----------
+    data_frame : pd.DataFrame
+        The DataFrame containing the data
+    mapping_required : dict
+        The required mapping from load_from_data_frame that helps identify the transaction_id column in the DataFrame
+    unmatched_transactions : list
+         A list of objects with the structure { transaction_id: instrument_identifiers }
+
+    Returns
+    -------
+    A filtered list of unmatched_transactions.
+    """
+    # Create a list var to hold the filtered transactions to be returned
+    filtered_unmatched_transactions = []
+
+    # Create a unique list of transaction_ids from the dataframe
+    valid_txids = data_frame[mapping_required.get("transaction_id")].unique()
+
+    # Iterate through the transactions and if the transaction_id is in the dataframe, add it to the list to be returned
+    for unmatched_transaction in unmatched_transactions:
```

Comment:
if `unmatched_transactions` is a unique list, you can use `set.difference`

Reply:
set operations should be significantly faster for operations that check the values within them (as opposed to iteration over its items where lists are faster)
This is a super important when we have portfolios with > 100,000s of txns

Reply:
I think i want Set.intersection() here, rather than difference. The values that are in the returned list and the acceptable list.

Reply:
Although as unmatched_transactions is a list of dictionaries, it will not be converted into a set. I think I can make the valid_txids a set but the for loop will have to stay.